File: Sources/Gantry/Services/TelegramService.swift
import Foundation

/// Outbound Telegram notifications through the official Bot API (https://core.telegram.org/bots/api).
/// No third-party gateway: a bot token + chat id are all it needs. Deliberately small and
/// dependency-free so the same shape ports 1:1 to the Windows (C#) and Linux (Python) clients.
///
/// Settings live in `AppSettings` under keys shared verbatim across platforms:
/// `telegram-enabled` / `telegram-bot-token` / `telegram-chat-id`. Telegram fires on the same events as
/// the native notifications (finished / error / paused / low filament / humidity), only when enabled and
/// configured. See docs/telegram.md.
enum TelegramService {
    /// Sends a message for one event when Telegram is enabled and configured. `title` + `body` mirror the
    /// native notification; the visible text is the shared `format` below. Main-actor because it reads
    /// `AppSettings.shared`; the actual network call hops off to a detached Task.
    @MainActor
    static func notify(printer: String, title: String, body: String) {
        let settings = AppSettings.shared
        guard settings.telegramEnabled else { return }
        if settings.telegramMuteUntil != nil { return }   // alerts silenced by /mute
        let token = settings.telegramBotToken.trimmingCharacters(in: .whitespacesAndNewlines)
        let chat = settings.telegramChatID.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !token.isEmpty, !chat.isEmpty else { return }
        let text = format(printer: printer, title: title, body: body)
        Task { await sendMessage(token: token, chatID: chat, text: text) }
    }

    /// The one message format every platform uses: `🖨 <printer> — <title>` then the body on a new line.
    static func format(printer: String, title: String, body: String) -> String {
        var text = "🖨 \(printer) — \(title)"
        if !body.isEmpty { text += "\n\(body)" }
        return text
    }

    /// Low-level send. Returns true on HTTP 200, so the Settings "Test" button can report success.
    @discardableResult
    static func sendMessage(token: String, chatID: String, text: String) async -> Bool {
        guard let url = URL(string: "https://api.telegram.org/bot\(token)/sendMessage") else { return false }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        var form = URLComponents()
        form.queryItems = [
            URLQueryItem(name: "chat_id", value: chatID),
            URLQueryItem(name: "text", value: text),
            URLQueryItem(name: "disable_web_page_preview", value: "true")
        ]
        request.httpBody = form.percentEncodedQuery?.data(using: .utf8)
        request.timeoutInterval = 15
        do {
            let (_, response) = try await URLSession.shared.data(for: request)
            return (response as? HTTPURLResponse)?.statusCode == 200
        } catch {
            return false
        }
    }
}

File: windows/Gantry.Windows/Services/TelegramService.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Threading.Tasks;

using System.Linq;

namespace Gantry.Services;

/// <summary>Outbound Telegram notifications through the official Bot API. The C# mirror of the macOS
/// TelegramService: same shared settings keys and the same message format. Fires on the same events as the
/// native notifications, only when enabled and not muted. See docs/telegram.md.</summary>
public static class TelegramService
{
    private static readonly HttpClient Http = new() { Timeout = TimeSpan.FromSeconds(15) };

    public static void Notify(string printer, string title, string body)
    {
        if (!AppSettings.TelegramEnabled) return;
        if (AppSettings.TelegramMuteUntil != null) return;   // silenced by /mute
        var token = AppSettings.TelegramBotToken.Trim();
        var chat = AppSettings.TelegramChatId.Trim();
        if (token.Length == 0 || chat.Length == 0) return;
        _ = SendMessageAsync(token, chat, Format(printer, title, body));
    }

    /// <summary>The one message format shared by every platform: "🖨 &lt;printer&gt; — &lt;title&gt;" then the body.</summary>
    public static string Format(string printer, string title, string body)
        => string.IsNullOrEmpty(body) ? $"🖨 {printer} — {title}" : $"🖨 {printer} — {title}\n{body}";

    /// The persistent bottom bar of commands, shared with the bot. Every outgoing message carries it, so
    /// it installs itself on the user's very first alert: the printer picker is an inline keyboard glued
    /// to one message, and once the chat scrolls past it there is no way back without scrolling up.
    public static string CommandKeyboard { get; } = System.Text.Json.JsonSerializer.Serialize(new
    {
        keyboard = new[]
        {
            new[] { "/status", "/all" }, new[] { "/spools", "/history" },
            new[] { "/watch 10m", "/mute 2h" }, new[] { "/help" }
        }.Select(r => r.Select(t => new { text = t }).ToArray()).ToArray(),
        resize_keyboard = true
    });

    public static async Task<bool> SendMessageAsync(string token, string chatId, string text,
                                                    string? replyMarkup = null)
    {
        try
        {
            var fields = new Dictionary<string, string>
            {
                ["chat_id"] = chatId, ["text"] = text, ["disable_web_page_preview"] = "true"
            };
            fields["reply_markup"] = replyMarkup ?? CommandKeyboard;
            var content = new FormUrlEncodedContent(fields);
            var response = await Http.PostAsync($"https://api.telegram.org/bot{token}/sendMessage", content);
            return response.IsSuccessStatusCode;
        }
        catch { return false; }
    }
}
